### ISSUE TITLE
Question: How to use s3backer with a GCS bucket?

### ISSUE THREAD

**Issue by Sunny-Dee** [archiecobbs/s3backer#208]
This is [technically possible](https://cloud.google.com/storage/docs/interoperability) but I am not able to get it to work. I created an [HMAC Key](https://cloud.google.com/storage/docs/authentication/managing-hmackeys) and saved it in this format `ACCESS_ID:SECRET` to `~/.s3backer_passwd` then ran: 
```shell
s3backer --blockCacheFile=blockcachefile --baseURL="https://storage.googleapis.com/" --authVersion="aws2" --blockSize=128k --size=200g --listBlocks my-bucket ~/fs/s3
```
And get the error:
```
ERROR: rec'd 403 response: HEAD https://storage.googleapis.com/my-bucket/00000000
s3backer: can't read data store meta-data: Operation not permitted
```
It's likely the base URL is wrong but [the docs](https://cloud.google.com/storage/docs/interoperability#xml_api) don't seem to point to anything else useful. Any help or examples would be greatly appreciated.

**Comment by archiecobbs** [comment]
No idea... did you also try with `--authVersion=aws4`?

**Comment by Sunny-Dee** [comment]
I did, unfortunately no luck. Thanks though!

**Comment by archiecobbs** [comment]
Yeah it's odd. I looked over the Google docs and nothing jumped out as a potential problem.

You could try running with the `-f` (run in foreground) and `--debug-http` (show HTTP traffic) flags.

It's possible that might reveal a clue.

Also, you could try creating a bucket that is public, and then mount it read-only (use `--readOnly`). This should then not require any authentication. This might tell us if the error is related to authentication or something else.

**Comment by Sunny-Dee** [comment]
Thank you! Running it with `-f` and `--debug-http`, it seems like it is the authentication header. The request has the AWS format, but based on the GCS docs it should have [this](https://cloud.google.com/storage/docs/aws-simple-migration#authentication) format. Here is the error I get (It retries a few times before exiting): 
 
```
*   Trying 173.194.203.207:443...
* Connected to storage.googleapis.com (173.194.203.207) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=storage.googleapis.com
*  start date: Aug  7 12:25:17 2023 GMT
*  expire date: Oct 30 12:25:16 2023 GMT
*  subjectAltName: host "storage.googleapis.com" matched cert's "storage.googleapis.com"
*  issuer: C=US; O=Google Trust Services LLC; CN=GTS CA 1C3
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x55804dfb0de0)
> HEAD /mybucket/00000000 HTTP/2
Host: storage.googleapis.com
user-agent: s3backer/2.0.2/2.0.2
accept: */*
date: Tue, 05 Sep 2023 17:18:55 GMT
authorization: AWS **MY_ACCESS_ID:MY_SECRET**

* Connection state changed (MAX_CONCURRENT_STREAMS == 100)!
< HTTP/2 404 
< x-guploader-uploadid: ***
< content-type: application/xml; charset=UTF-8
< content-length: 186
< date: Tue, 05 Sep 2023 17:20:13 GMT
< expires: Tue, 05 Sep 2023 17:20:13 GMT
< cache-control: private, max-age=0
< server: UploadServer
< 
* Connection #0 to host storage.googleapis.com left intact
s3backer: auto-detection failed; using configured block size 128k and file size 200g

```

Based on this, I won't be able to use the tool unless there is a code change [here](https://github.com/archiecobbs/s3backer/blob/master/http_io.c#L2518), maybe?

**Comment by archiecobbs** [comment]
Can you conduct that same experiment with `--authVersion=aws4` and show the same debug headers, etc.?

From my reading, the Google docs seem to imply that `aws4` is the only supported authentication method.

**Comment by Sunny-Dee** [comment]
I ran it with `aws4` and the authorization format looks correct now, but I get the same type of error.
```
s3backer: auto-detecting block size and total file size...
*   Trying 74.125.197.207:443...
* Connected to storage.googleapis.com (74.125.197.207) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=storage.googleapis.com
*  start date: Aug  7 12:25:17 2023 GMT
*  expire date: Oct 30 12:25:16 2023 GMT
*  subjectAltName: host "storage.googleapis.com" matched cert's "storage.googleapis.com"
*  issuer: C=US; O=Google Trust Services LLC; CN=GTS CA 1C3
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x55a03bfadde0)
> HEAD /my-bucket/00000000 HTTP/2
Host: storage.googleapis.com
user-agent: s3backer/2.0.2/2.0.2
accept: */*
x-amz-date: 20230905T212715Z
x-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855
authorization: AWS4-HMAC-SHA256 Credential=**MY_ACCESS_ID**/us-east-1/s3/aws4_request, SignedHeaders=host;x-amz-content-sha256;x-amz-date, Signature=**HASHED_SECRET**

## Retries a few times. Below is the last entry:

* Connection #0 to host storage.googleapis.com left intact
2023-09-05 21:27:42 ERROR: rec'd 400 response: PUT https://storage.googleapis.com/my-bucket/s3backer-mounted
2023-09-05 21:27:47 INFO: retrying query (attempt #9): PUT https://storage.googleapis.com/my-bucket/s3backer-mounted
*   Trying 142.250.107.207:443...
* Connected to storage.googleapis.com (142.250.107.207) port 443 (#0)
* ALPN, offering h2
* ALPN, offering http/1.1
* successfully set certificate verify locations:
*  CAfile: /etc/ssl/certs/ca-certificates.crt
*  CApath: /etc/ssl/certs
* SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
* ALPN, server accepted to use h2
* Server certificate:
*  subject: CN=storage.googleapis.com
*  start date: Aug  7 12:25:17 2023 GMT
*  expire date: Oct 30 12:25:16 2023 GMT
*  subjectAltName: host "storage.googleapis.com" matched cert's "storage.googleapis.com"
*  issuer: C=US; O=Google Trust Services LLC; CN=GTS CA 1C3
*  SSL certificate verify ok.
* Using HTTP2, server supports multi-use
* Connection state changed (HTTP/2 confirmed)
* Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
* Using Stream ID: 1 (easy handle 0x55a03bfd6520)
> PUT /my-bucket/s3backer-mounted HTTP/2
Host: storage.googleapis.com
user-agent: s3backer/2.0.2/2.0.2
accept: */*
x-amz-date: 20230905T212716Z
content-type: text/plain
content-md5: jWgufICkacha8QU5HXy/SA==
x-amz-meta-s3backer-mount-token: 290515c9
x-amz-acl: private
x-amz-content-sha256: 0eeb76e244428ecf820774c52e0728ed5c0e75b2632081e7a767fe9b4d4d0493
authorization: AWS4-HMAC-SHA256 Credential=**MY_ACCESS_ID**/20230905/us-east-1/s3/aws4_request, SignedHeaders=content-md5;content-type;host;x-amz-acl;x-amz-content-sha256;x-amz-date;x-amz-meta-s3backer-mount-token, Signature=**HASHED_SECRET**
content-length: 35

* Connection state changed (MAX_CONCURRENT_STREAMS == 100)!
* We are completely uploaded and fine
< HTTP/2 400 
< x-guploader-uploadid: ADPycdsdxBpNK2U4JIvkQFR0hdC8B0dQazN0fVJCN1OREdMlOxGP2zdgE8M2aP6cGKUxuIPItZ5_-2wikSuJDbCEczEidnybvKza
< content-type: application/xml; charset=UTF-8
< content-length: 298
< vary: Origin
< date: Tue, 05 Sep 2023 21:27:47 GMT
< server: UploadServer
< 
* Connection #0 to host storage.googleapis.com left intact
2023-09-05 21:27:47 ERROR: rec'd 400 response: PUT https://storage.googleapis.com/my-bucket/s3backer-mounted
2023-09-05 21:27:47 ERROR: giving up on: PUT https://storage.googleapis.com/my-bucket/s3backer-mounted
2023-09-05 21:27:47 ERROR: error reading mount token on https://storage.googleapis.com/my-bucket/: Input/output error
s3backer: error creating s3backer_store: Input/output error
```
It is possible the signature part of the authorization is not derived appropriately? It's hard for me to tell, but it should be derived  from the secret like [this](https://cloud.google.com/storage/docs/authentication/signatures#derive-key).

Thanks so much for looking into it!

**Comment by archiecobbs** [comment]
> It is possible the signature part of the authorization is not derived appropriately?

Yes. s3backer derives appropriately for Amazon AWS, so Google must be expecting something different... even though they also seem to be saying that it should be compatible. So I don't know what the problem is.

Here's what they say:

> Migrate from Amazon S3 to Cloud Storage in a simple migration scenario
> 
> In order to make requests to Cloud Storage, you need to complete the following steps:
> 
> * Set a default Google project.
> * Get an HMAC (hash-based message authentication code) key.
> 
> * In your existing tools or libraries, make the following changes:
>   * Change the request endpoint to use the Cloud Storage XML API request endpoint.
>   * Replace the Amazon Web Services (AWS) access and secret key with the corresponding Cloud Storage access ID and secret (collectively called your Cloud Storage HMAC key).
> 
>   * Make sure your x-amz- headers use supported Cloud Storage values. For example, x-amz-storage-class should use one of the available Cloud Storage storage classes.
> 
>   * When you use the Cloud Storage XML API in a simple migration scenario, specifying the AWS signature identifier in the Authorization header lets Cloud Storage know to expect x-amz-* headers and Amazon S3 ACL XML syntax in your request. Cloud Storage processes x-amz-* headers that have an x-goog-* equivalent, such as those listed in the headers table.

Maybe they don't like one of the `x-amz-*` headers we're sending?

It's possible Google support or one of their user groups may be able to help.

**Comment by Sunny-Dee** [comment]
Thank you!

Turns out it was a configuration issue on the bucket side. I initially set the bucket's access control to be uniform, using IAM policies, but it needed to be Fine-grained which allows for Access Control Lists (ACLs). From the [docs](https://cloud.google.com/storage/docs/access-control/?_ga=2.199202479.-1836013100.1688573347&_gac=1.224925032.1692396227.Cj0KCQjwrfymBhCTARIsADXTabnY_4P2LCsKdXU56MfLrGkSfcfIXDKZ1HoAOMW_uB72DA17bF6fxbkaAh29EALw_wcB) 
> ACLs are a legacy access control system for Cloud Storage designed for interoperability with Amazon S3

I was able to start s3backer with the following command:
```shell
s3backer --blockSize=128k --size=1G --listBlocks s3backer-bucket --baseURL="https://storage.googleapis.com/" --authVersion="aws4"  --region="us"  ~/s3b/mnt/
```

I see the `s3backer-mounted` object in the bucket. However, for some reason, formatting the file fails. I am using: 
```shell
sudo mkfs.ext4 -E nodiscard -F ~/s3b/test/file
```

and get the following error:
```
mke2fs 1.46.2 (28-Feb-2021)
Warning: could not erase sector 2: Operation not permitted
Creating filesystem with 262144 4k blocks and 65536 inodes
Superblock backups stored on blocks: 
	32768, 98304, 163840, 229376

Allocating group tables: done                            
Warning: could not read block 0: Operation not permitted
Warning: could not erase sector 0: Operation not permitted
Writing inode tables: done                            
Creating journal (8192 blocks): done
Writing superblocks and filesystem accounting information: mkfs.ext4: Operation not permitted while writing out and closing file system
```
Each time I try, an object gets written to the bucket. I tried the same steps with a public s3 bucket and it succeeds. Thanks for all the support.

**Comment by archiecobbs** [comment]
Sounds like maybe it has permission to write objects but not delete them?

If so you should see this in the log... PUT's returning 200 but DELETE's returning 403.

**Comment by Sunny-Dee** [comment]
It's a permissions thing but not sure exactly what. When starting s3backer, `GET` requests succeed:
```
2023-09-18 20:59:22 DEBUG: success: GET https://s3backer-v2.storage.googleapis.com/?marker=00001dff&max-keys=1000
```

But when I try to format the file system, the tool logs a bunch of errors trying to get objects from the bucket like this:
```
< HTTP/2 403 
< x-guploader-uploadid: ADPycdsnbiJBqoYLlQJSLtqh7Wg6wuUYUi0chzQdi5HIu9v61GqWkAEDih1p9tapgjdPHWgEXk6e9DySNLp0TjFcjwda1zEITV0q
< content-type: application/xml; charset=UTF-8
< content-length: 779
< date: Mon, 18 Sep 2023 21:00:00 GMT
< expires: Mon, 18 Sep 2023 21:00:00 GMT
< cache-control: private, max-age=0
< server: UploadServer
< 
* Connection #0 to host s3backer-v2.storage.googleapis.com left intact
2023-09-18 21:00:00 ERROR: rec'd 403 response: GET https://s3backer-v2.storage.googleapis.com/00000084
2023-09-18 21:00:00 DEBUG: HTTP 403 status response payload:
<?xml version='1.0' encoding='UTF-8'?><Error><Code>SignatureDoesNotMatch</Code><Message>Access denied.</Message><Details>The request signature we calculated does not match the signature you provided. Check your Google secret key and signing method.</Details><StringToSign>AWS4-HMAC-SHA256
20230918T210000Z
20230918/us/s3/aws4_request
612527d4aad7ba4aaca4637931b2ac2b6c81a6c616d664638b0efc514dfbb41a</StringToSign><CanonicalRequest>GET
/00000084
```

**Comment by archiecobbs** [comment]
Ok this is good that they show the `StringToSign` because now we can compare.

But I'm skeptical that's actually the problem, because otherwise the first `GET` would have also failed you'd think... Or maybe it succeeded because you allow public listing but not public object reads?

Anyway, edit `http_io.c` and update `#define DEBUG_AUTHENTICATION 1` , rebuild, and try again with that same logging.

This will add a bunch of logging related to authentication. Note: your secret key will appear in the logged output so elide that if you cut & paste.

Then let's compare our `StringToSign` with theirs.

**Comment by Sunny-Dee** [comment]
Thank you so much, looking at the `GET` request that succeeded versus the one that failed, the only difference was the `accept-encoding` header. I disabled it [here](https://github.com/archiecobbs/s3backer/blob/master/http_io.c#L1560), rebuilt and tried again. Now we are good to go! 

Not sure what about the logic sets the header differently, but disabling it altogether worked for me. 

Thank you for the support!

**Comment by archiecobbs** [comment]
Ah! We should not be sending an empty `Accept-Encoding` header, which is what would happen if neither compression nor encryption was enabled.

That bug should be fixed in 0e07944.
